libsmtutil/CHCSmtLib2Interface.cpp

```diff
@@ -223,3 +225,275 @@ std::string CHCSmtLib2Interface::createHeaderAndDeclarations() {
+							}
+							for (size_t i = 1; i < _subExpr.size(); i++)
+								arguments.emplace_back(toSMTUtilExpression(_subExpr[i]));
+							if (auto tupleSort = lookupKnownTupleSort(op); tupleSort)
+							{
+								auto sortSort = std::make_shared<SortSort>(tupleSort.value());
+								return Expression::tuple_constructor(Expression(sortSort), arguments);
+							}
+							if (m_knownVariables.find(op) != m_knownVariables.end())
+							{
+								return smtutil::Expression(op, std::move(arguments), m_knownVariables.at(op));
+							}
+							else {
+								std::set<std::string> boolOperators{"and", "or", "not", "=", "<", ">", "<=", ">=",
+																	"=>"};
+								sort = contains(boolOperators, op) ? SortProvider::boolSort : arguments.back().sort;
+								return smtutil::Expression(op, std::move(arguments), std::move(sort));
+							}
+							smtAssert(false, "Unhandled case in expression conversion");
+						} else {
+							// check for const array
+							if (_subExpr.size() == 2 and !isAtom(_subExpr[0]))
+							{
+								auto const& typeArgs = asSubExpressions(_subExpr.front());
+								if (typeArgs.size() == 3 && typeArgs[0].toString() == "as" && typeArgs[1].toString() == "const")
+								{
+									auto arraySort = toSort(typeArgs[2]);
+									auto sortSort = std::make_shared<SortSort>(arraySort);
+									return smtutil::Expression::const_array(Expression(sortSort), toSMTUtilExpression(_subExpr[1]));
+								}
+								if (typeArgs.size() == 3 && typeArgs[0].toString() == "_" && typeArgs[1].toString() == "int2bv")
+								{
+									auto bvSort = std::dynamic_pointer_cast<BitVectorSort>(toSort(_subExpr[0]));
+									smtAssert(bvSort);
+									return smtutil::Expression::int2bv(toSMTUtilExpression(_subExpr[1]), bvSort->size);
+								}
+								if (typeArgs.size() == 4 && typeArgs[0].toString() == "_")
+								{
+									if (typeArgs[1].toString() == "extract")
+									{
+										return smtutil::Expression(
+												"extract",
+												{toSMTUtilExpression(typeArgs[2]), toSMTUtilExpression(typeArgs[3])},
+												SortProvider::bitVectorSort // TODO: Compute bit size properly?
+												);
+									}
+								}
+							}
+
+							smtAssert(false, "Unhandled case in expression conversion");
+						}
+					}
+			}, _expr.data);
+		}
+	};
+
+
+#define precondition(CONDITION) if (!(CONDITION)) return {}
+std::optional<smtutil::Expression> CHCSmtLib2Interface::invariantsFromSolverResponse(std::string const& _response) const
+{
+	std::stringstream ss(_response);
+	std::string answer;
+	ss >> answer;
+	precondition(answer == "sat");
+	SMTLib2Parser parser(ss);
+	precondition(!parser.isEOF()); // There has to be a model
+	std::vector<SMTLib2Expression> parsedOutput;
+	try
+	{
+		while (!parser.isEOF())
+			parsedOutput.push_back(parser.parseExpression());
+	}
+	catch(SMTLib2Parser::ParsingException&)
+	{
+		return {};
+	}
+	precondition(parser.isEOF());
```

Comment:
Here I am just checking that I have really parsed everything and I am at the end of the stream.
It seems this check is superfluous, I can most likely just remove it, because, as you say, either I exited the while loop, so I have to be at the end, or there was a parsing exception and the method already returned.

But, beyond this check, I have to process what I have parsed.

I think, I will just remove this check, do you agree?

Reply:
Ah OK, `#define precondition(CONDITION) if (!(CONDITION)) return {}` negates the condition and only returns then, so the section below is indeed not dead code. In that case, you can remove the check, or better yet, replace it with an `smtAssert`, since this should never really happen.

Reply:
Replaced with `smtAssert`!